Add space if qop is first param

vasilakisfil opened the pull request:
Hi,

I just found a small bug, that when qop appears first in AuthHeader map params, it the serialization of it (through Display trait) will not be correct, specifically it doesn't add a space after the scheme.

Apart from that I haven't really figured out when the second case is used (`} else if index == 0 && key == &"auth".to_string() {` (line 36) is used, by looking in the relevant RFCs, I didn't find any info, regarding a key with name `auth`, but left is just to make sure I don't break anything. However, I think we need to add space there as well.

Finally I have found out that using regular HashMap is a bit tricky to do testing and reason things, since output is not deterministic but random, so I would suggest to to use a BTreeMap instead. Sure performance is a bit slower, but in practice different will be very small if any, and testing will be much easier :)

bytebuddha commented:
Thanks for the work, and you are correct in your assumption about the the second case, it should also include a space, do you think you could fix this one as well?

I am OK with using BTreeMap instead HashMap in AuthHeader ill create an issue for that to be done separately

vasilakisfil commented:
Hi, yes I can fix that, but about which assumption am I right?  :P that there is no `auth` key found in any rfc, or that we also need to add a space in case of an `auth` key ?

vasilakisfil commented:
I am wondering if you meant to write instead of `auth`, `nc`. This parameter, needs to be formatted using the `{0:8}` precision.

bytebuddha commented:
Upon rereading the relevant RFC you are indeed correct it should be the nonce-count field (nc).

vasilakisfil commented:
Hi finally I found some time to fix that! I renamed the "auth" to "qc" but I also noticed that adding leading zeros to a String format param doesn't make sense (specifically it only adds trailing spaces), so removed that and moved the responsibility to the library user to make sure to have the appropriate leading zeros (probably using the `{:08}` directive) before converting that to String and passing it to libsip Auth header.

Let me know if you agree with this change. The alternative it to do the transformation from String -> Integer inside the Display trait but I don't feel like it's a good idea to fail there for that reason.

bytebuddha commented:
Look's great, I think your correct about not failing in the Display trait, that should only fail if it is unable to render
